### ISSUE TITLE
Consider using RaggedTensors

### ISSUE THREAD

**Issue by tabergma** [RasaHQ/rasa#6909]
**Description of Problem**: Currently, we are padding all batches to the same shape in `RasaModelData.prepare_batch`. For our dense data we could use `RaggedTensors` ([doc](https://www.tensorflow.org/guide/ragged_tensor)) instead of padding the data ourself. This should simplify the code. However, `RaggedTensors` cannot be used with `tf.data.Dataset.from_generator` right now. It seems that this was already tackled in [https://github.com/tensorflow/tensorflow/pull/41981](https://github.com/tensorflow/tensorflow/pull/41981), but not yet released.

**Overview of the Solution**: Once `RaggedTensors` can be used with `tf.data.Dataset.from_generator` update the code to use `RaggedTensors` and do not pad the data anymore. We might need to update some of our layers.

**Comment by Ghostvv** [comment]
I remember we also wanted to redo `tf.data.Dataset.from_generator` anyhow

**Comment by lithuak** [comment]
@tabergma Hey, the support for `RaggedTensor` in `tf.data.Dataset.from_generator` was released quite a while ago I believe. Actually we've put the example of `RaggedTensor` usage into the docstring for `tf.data.Dataset.from_generator`, check it out :)

**Comment by koernerfelicia** [comment]
@dakshvar22 have you looked into this? Should it be moved to in progress, or is it good where it is on the board?

**Comment by dakshvar22** [comment]
Haven't made progress on it. Wanted to start on it once we complete the transition to Keras API.

**Comment by dakshvar22** [comment]
Update: Unable to make ragged tensors currently work with our setup, i.e. a custom keras model + data generator of type `tf.keras.utils.Sequence`. Could be hitting a bug for which I've created an [issue](https://github.com/tensorflow/tensorflow/issues/48188) on tensorflow repo for now. 

@lithuak Looks like you had pushed [support](https://github.com/tensorflow/tensorflow/pull/41981) for ragged tensors inside `tf.data.Dataset.from_generator`. While our setting is not exactly the same, but if you have any ideas on what could unblock us in the above linked issue, they are more than welcome!
